app/src/main/java/com/example/tasky/core/data/remote/BaseRepository.kt

```diff
@@ -0,0 +1,53 @@
+package com.example.tasky.core.data.remote
+
+import com.example.tasky.core.data.Resource
+import com.squareup.moshi.Moshi
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import okhttp3.ResponseBody
+import retrofit2.HttpException
+import retrofit2.Response
+import java.io.IOException
+
+abstract class BaseRepository {
+
+    suspend fun <T> safeApiCall(apiToBeCalled: suspend () -> Response<T>): Resource<T> {
+        return withContext(Dispatchers.IO) {
+            try {
+                val response: Response<T> = apiToBeCalled()
+
+                if (response.isSuccessful) {
+                    response.body()?.let {
+                        Resource.Success(data = it)
+                    } ?: Resource.Error(errorMessage = "Null body")
+
+                } else {
+                    val errorResponse: ErrorResponse? = convertErrorBody(response.errorBody())
+                    Resource.Error(errorMessage = errorResponse?.message ?: "Something went wrong")
+                }
+
+            } catch (e: HttpException) {
+                Resource.Error(errorMessage = e.message ?: "Something went wrong")
+            } catch (e: IOException) {
+                Resource.Error("Please check your network connection")
+            } catch (e: Exception) {
+                Resource.Error(errorMessage = "Something went wrong")
+            }
```

Comment:
Be careful with catching general exceptions inside suspend functions. You can break coroutine cancellation that way. If a coroutine is cancelled, the currently suspending function (here, it would be `apiToBeCalled`) will throw a CancellationException. If your catch block eats that up, this exception won't be properly propagated up the coroutine scopes which can cause some coroutines to run longer than they should.
You can fix this by adding this:

```
if(e is CancellationException) throw e
```
or alternatively by only catching specific exceptions and not CancellationExceptions.

In this video I show the issue with that a bit more in detail (starting at 17:45):
https://www.youtube.com/watch?v=VWlwkqmTLHc